### PR TITLE
feat(react-components): Generalizing code so it is ready for onboarding annotations

### PR DESCRIPTION
**Pull request by nilscognite**
#### Type of change
Feat  
https://cognitedata.atlassian.net/browse/

## Description :pencil:

**BaseCommand** Add mac support

**BaseEditTool:** Small adjustment

**BaseFilterCommand:** Generalizin the invocation

**BaseTool**: add onDeleteKey and onEscapeKey ++

**RenderTargetCommand:** Add local convenience func

**MockFilterCommand :** Made it simpler

**BaseFilterItemCommand:** Implement the `invokeCore`, so the sub classes do not need to override this.

**PointCloudFilterCommand:** Remove the `invokeCore`

**DomainObject:** Change signature of notify, so it can be simpler from the outside.

**RootDomainObject:** Move stk pointer to this calss from RenderTarget. 

**VisualDomainObject:** Let render target be optional, because it is only one.

**BaseCreator:** Renaming + new virtual function

**BaseDragger:** Generalized concept of constrains (shift key), used by the sub classes (had to use the unit system in the class)

**ChangedDescription:** Moved it out of DomainObjectChange.ts

**Changes:** Add several new changes, for updating only a part of the domain object.

**DomainObjectChange:** Little cleanup + added `getChangedDescriptionByType`

**DomainObjectPanelUpdater:** Added dragging change. This is when the object is dragged. This is done several places. So dragging is the same as geometry, except that dragging  is using during dragging and geometry is when the dragged has finished or all other cases.

**CommandsController:** Return true if anything changes in some of the methods + add  setActiveToolByType method

**RevealRenderTarget**: Move sdk to RootDomainObject + add setDefaultTool function

**BoxUtils, CylinderUtils, PrimitiveUtils:** Merged functions from Fusion/React-comp to render wireframe. Reused in several classes.

**arrayExtensions:** Let remove return true when the object is removed.

**mathExtensions:** Add `forceAngleAround0`

**stringExtension:**  `equalsIgnoreCaseAndSpace`: Fixed a bug

**vectorExtensions:** Make Vector2 as function parameter when possible

**Range3:** Added `getCircleRangeMargin`, port from optimizer. Used to find bounding box for cylinder.

**getCorners:** Generator for corners of a box

**isPointVisibleByPlanes:** Added some utility functions for the planes. Used in the views.

**primitives** Generalizing Cylinder/Box as a primitive. This is used by BoxDomainObject and annotations. The rest is taken from fusion

Renamed `BoxRenderStyle `to `SolidPrimitiveRenderStyle`, used by both  boxes and cylinders

**getIconByPrimitiveType:** Added icon for cylinder. This is not correct, waiting for the correct icon.

**ObservationsView:** Removed function, reused another similar

**PrimitiveEditTool:** Smaller refactoring

**PrimitiveType:** Added cylinder

**SolidDomainObject :** Common base class for Cylinder/Box domain object

**box/*:** Changed base class for the domain object.  Change to use Euler angles instead of just `zRotation`. In the view and the dragger, this change a lot. Let the dragger and view use all euler angles. In the view, some code are used from the part of Annotation. These functions is now in `BoxUtils `and `PrimitiveUtils`.

**cylinder/*:** Added a `CylinderDomainObject `with dragger, view and creator. Equal pattern as the box domain object.
 
 
**Basically the rest of the changes are related to changes in the base object, use new utility function etc.**

## How has this been tested? :mag:

This is just refactoring, nothing in the UI is changed. Use measurement with box and see if it is working fine. You can also check with the crop box

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [x] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: react-components/src/architecture/concrete/primitives/box/BoxDragger.ts [hunk 1]
```diff
@@ -241,52 +220,85 @@ export class BoxDragger extends BaseDragger {
...
+      let oldAngle;
+      if (this._face.index === 0) {
+        oldAngle = originalBox.rotation.x;
+      } else if (this._face.index === 1) {
+        oldAngle = originalBox.rotation.y;
+      } else if (this._face.index === 2) {
+        oldAngle = originalBox.rotation.z;
+      } else {
+        return false;
+      }
```
Comment (haakonflatval-cognite) [inline react-components/src/architecture/concrete/primitives/box/BoxDragger.ts:275]: I must admit that I don't understand what the `shift` boolean does here :sweat_smile: I assume it says whether the shift button is pressed or not, but I don't understand its effect. Could this part be put in a separate named function to make it more clear what's happening?

File: react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    point: Vector3,
+    pickInfo: PrimitivePickInfo
+  ): void {
+    if (domainObject instanceof BoxDomainObject) {
+      PrimitiveEditTool.setCursorForBox(tool, domainObject, point, pickInfo);
+    } else if (domainObject instanceof CylinderDomainObject) {
+      PrimitiveEditTool.setCursorForCylinder(tool, domainObject, point, pickInfo);
+    }
+  }
+
```
Comment (haakonflatval-cognite) [inline react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts:347]: I suppose you've already thought about this - but isn't it a bit unfortunate that the `PrimitiveEditTool` must know about the different primitives?

File: react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+        const { _creator: creator } = this;
+        if (!creator.preferIntersection) {
+          // Hover in the "air"
+          const ray = this.getRay(event);
+          if (creator.addPoint(ray, undefined, true)) {
+            this.setDefaultCursor();
+            return;
+          }
+        }
+        const intersection = await this.getIntersection(event);
```
Comment (haakonflatval-cognite) [inline react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts:93]: This code is very hard to follow - would it be easier to make it clear if it was split into functions?

**Review by haakonflatval-cognite — COMMENTED** [review]
Looks good! Just wrote a few comments I'd like to have addressed, but nothing critical

File: react-components/src/architecture/base/commands/BaseFilterCommand.ts [hunk 4]
```diff
@@ -134,6 +134,14 @@ export abstract class BaseFilterCommand extends RenderTargetCommand {
 }
 
 export abstract class BaseFilterItemCommand extends RenderTargetCommand {
+  // ==================================================
+  // OVERRIDES
+  // ==================================================
+
+  public override invokeCore(): boolean {
+    return this.setCheckedCore(!this.isChecked);
+  }
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/commands/BaseFilterCommand.ts:143]: No. I made a comment about it

[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/commands/BaseFilterCommand.ts:143]:
```
  public override invokeCore(): boolean {
    // Toggle the checked state, you do npt need to override this method
    // as long as setCheckedCore and isChecked are overridden
    return this.setCheckedCore(!this.isChecked);
  }

```

File: react-components/src/architecture/base/utilities/extensions/mathExtensions.ts
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/utilities/extensions/mathExtensions.ts:1]: I will make a separate PR for this. Please remind me

File: react-components/src/architecture/base/utilities/geometry/isPointVisibleByPlanes.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    if (!isPointVisibleByPlanes(planes, corner)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// TODO: This function is defined in Reveal. Reuse that instead?
+export function isPointVisibleByPlanes(planes: Plane[], point: Vector3): boolean {
+  return planes.every((p) => p.distanceToPoint(point) >= 0);
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/utilities/geometry/isPointVisibleByPlanes.ts:36]: I disagree in both. 
1) Always reuse.
2) **Visible** is used all other places in Tree.js, and in Reveal. Keep naming consist is important for me and most other developers.

Yes. I will keep the TODO

File: react-components/src/architecture/base/utilities/primitives/BoxUtils.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  new Vector3(-HALF_SIDE, -HALF_SIDE, -HALF_SIDE),
+  new Vector3(+HALF_SIDE, -HALF_SIDE, -HALF_SIDE),
+  new Vector3(+HALF_SIDE, +HALF_SIDE, -HALF_SIDE),
+  new Vector3(-HALF_SIDE, +HALF_SIDE, -HALF_SIDE),
+  new Vector3(-HALF_SIDE, -HALF_SIDE, +HALF_SIDE),
+  new Vector3(+HALF_SIDE, -HALF_SIDE, +HALF_SIDE),
+  new Vector3(+HALF_SIDE, +HALF_SIDE, +HALF_SIDE),
+  new Vector3(-HALF_SIDE, +HALF_SIDE, +HALF_SIDE)
+];
+export class BoxUtils {
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/utilities/primitives/BoxUtils.ts:22]: Ok, done, A little bit easier

File: react-components/src/architecture/base/utilities/sprites/createSprite.ts [hunk 7]
```diff
@@ -94,8 +94,8 @@ function createTexture(canvas: HTMLCanvasElement): Texture {
   texture.magFilter = NearestFilter;
 
   // texture.generateMipmaps = true; // Default
-  // texture.wrapS = THREE.ClampToEdgeWrapping;
-  // texture.wrapT = THREE.ClampToEdgeWrapping;
+  // texture.wrapS = ClampToEdgeWrapping;
+  // texture.wrapT = ClampToEdgeWrapping;
   texture.needsUpdate = true;
   return texture;
 }
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/utilities/sprites/createSprite.ts:101]: Removed

File: react-components/src/architecture/concrete/primitives/box/BoxDragger.ts (hunk 1 above)
Comment (nilscognite) [reply react-components/src/architecture/concrete/primitives/box/BoxDragger.ts:275]: Renamed to isShiftPressed. Also made everything clearer.

File: react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts (hunk 2 above)
Comment (nilscognite) [reply react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts:347]: You are correct. This is really bad.

File: react-components/src/architecture/base/utilities/primitives/BoxUtils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/utilities/primitives/BoxUtils.ts:22]: I hade to change the ordering of getCorners() to follow the same scheme as I had

File: react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts (hunk 3 above)
Comment (nilscognite) [reply react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts:93]: Yes it is. I want to want until the annotation is merge. Because I don't  know if everything here is actual wanted

File: react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts (hunk 2 above)
Comment (nilscognite) [reply react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts:347]: This made quite a bit of restructuring. I have made a virtual function called getEditToolCursor for getting the cursor while in a edit tool. Also setFocusInteractlve() is made virtual. So now, we don't have to check for specific objects in the PrimitiveEditTool.

Comment (nilscognite) [reply react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts]: Also made some matrix transformation simpler by making toViewerMatrix, fromViewerMatrixin the renderTarget. By this, I have removed all CDF_TO_VIEWER_TRANSFORMATION.clone().invert() in the code, since it is stored by the renderTarget instead.

File: react-components/src/architecture/base/utilities/geometry/isPointVisibleByPlanes.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (haakonflatval-cognite) [reply react-components/src/architecture/base/utilities/geometry/isPointVisibleByPlanes.ts:36]: I think we use `clipped` in Reveal - but whatever, I'll let it pass

**Review by haakonflatval-cognite — APPROVED** [review]
LGTM :+1:

File: react-components/src/architecture/base/utilities/primitives/BoxUtils.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  }
+}
+
+function createCubeCorners(): Vector3[] {
+  const box = new Box3(new Vector3().setScalar(-HALF_SIDE), new Vector3().setScalar(HALF_SIDE));
+  const result: Vector3[] = [];
+  for (const corner of getCorners(box)) {
+    result.push(corner.clone());
+  }
+  return result;
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/utilities/primitives/BoxUtils.ts:64]: Ha ha. Tried that first actually. Then you will end up with the same array

[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/utilities/primitives/BoxUtils.ts:64]: To make it fast getCorners() alays return the same vector3, in a mutated form

File: react-components/src/architecture/base/renderTarget/RevealRenderTarget.ts [hunk 9]
```diff
@@ -97,6 +99,14 @@ export class RevealRenderTarget {
     return this._viewer.domElement;
   }
 
+  public get toViewerMatrix(): Matrix4 {
+    return CDF_TO_VIEWER_TRANSFORMATION;
+  }
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/renderTarget/RevealRenderTarget.ts:104]: No. It makes it slow. It is so easy to see if someone do it.

[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/architecture/base/renderTarget/RevealRenderTarget.ts:104]: Bute I did another trick:

```
  public readonly toViewerMatrix = CDF_TO_VIEWER_TRANSFORMATION.clone();
  public readonly fromViewerMatrix = CDF_TO_VIEWER_TRANSFORMATION.clone().invert();
```

File: react-components/src/architecture/base/utilities/primitives/BoxUtils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (haakonflatval-cognite) [reply react-components/src/architecture/base/utilities/primitives/BoxUtils.ts:64]: Aaah, I see :see_no_evil:
